Introducing payment intent status constants

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This is another simple refactor PR introducing a new constant class to replace all references to payment intent status names. This is useful to centralize information, make it easier to reference the statuses, and later build specific logic for groups of statuses.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

Code review should be enough. Check if the tests are still passing. No behavior should change with this PR.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Introducing payment intent status constants

wjrosa committed:
Merge branch 'develop' into dev/introducing-payment-intent-constants

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Add missing include

wjrosa committed:
Grouping some statuses

wjrosa committed:
Grouping some statuses

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -889,21 +882,21 @@ private function process_payment_with_deferred_intent( int $order_id ) {
 				$redirect = $return_url;
-			}
 
-			if ( $payment_needed && ! in_array( $payment_intent->status, [ 'requires_confirmation', 'requires_action' ], true ) ) {
```
I just simplified this since both blocks contain similar conditions (`in_array( $payment_intent->status, [ 'requires_confirmation', 'requires_action' ], true )`)

mattallan commented on includes/class-wc-gateway-stripe.php:
```diff
@@ -888,17 +888,17 @@ public function verify_intent_after_checkout( $order ) {
 		}
 
-		if ( 'setup_intent' === $intent->object && 'succeeded' === $intent->status ) {
+		if ( 'setup_intent' === $intent->object && WC_Stripe_Payment_Intent_Status::SUCCESSFUL_STATUSES === $intent->status ) {
```
This line isn't correct. It's using `WC_Stripe_Payment_Intent_Status::SUCCESSFUL_STATUSES` which is comparing an array to a string.

wjrosa replied:
Good catch, Matt! Thanks for pointing it out. Fixed it

mattallan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -41,13 +41,6 @@ class WC_Stripe_UPE_Payment_Gateway extends WC_Gateway_Stripe {
-	 *
-	 * @type array
-	 */
-	const SUCCESSFUL_INTENT_STATUS = [ 'succeeded', 'requires_capture', 'processing' ];
```
Since this was publicly available, do we need to deprecate this before removing it? Anyone attempting to use `WC_Stripe_UPE_Payment_Gateway::SUCCESSFUL_INTENT_STATUS` will throw a PHP fatal error after upgrading.

wjrosa replied:
Good point. Deprecated it instead of removing it in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3648/commits/9f3895ac742aa8f0b23b36a3e7a63a7f3f0c09b1

mattallan reviewed (changes requested):
Nice work @wjrosa on these changes 💯 

I've left a couple of comments but something I was also curious about is whether we should rename this new class from `WC_Stripe_Payment_Intent_Status` to just `WC_Stripe_Intent_Status` as these constants are being used for both setup and payment intents?

wjrosa committed:
Merge branch 'develop' into dev/introducing-payment-intent-constants

wjrosa committed:
Fix incorrect string replacement

wjrosa committed:
Renaming constant class to WC_Stripe_Intent_Status

wjrosa committed:
Deprecating constant instead of removing it

wjrosa commented:
> Nice work @wjrosa on these changes 💯
> 
> I've left a couple of comments but something I was also curious about is whether we should rename this new class from `WC_Stripe_Payment_Intent_Status` to just `WC_Stripe_Intent_Status` as these constants are being used for both setup and payment intents?

Thank you for your review, Matt! I have addressed all of your points. Also renamed the class as suggested 👍

wjrosa committed:
Adding official doc link

mattallan reviewed (approved):
Thanks @wjrosa, i've checked the latest changes and believe this is good to be merged :)

wjrosa committed:
Merge branch 'develop' into dev/introducing-payment-intent-constants
